src/types/ErrorRate.ts

```diff
@@ -0,0 +1,167 @@
+import { NA, RequestHealth } from './Health';
+import { RateHealthConfig, ToleranceConfig } from './ServerConfig';
+import { serverConfig } from '../config/ServerConfig';
+import { ThresholdStatus, getRequestErrorsStatus, RATIO_NA } from './Health';
+
+export interface ErrorRatio {
+  global: { status: ThresholdStatus; protocol?: string };
+  inbound: { status: ThresholdStatus; protocol?: string };
+  outbound: { status: ThresholdStatus; protocol?: string };
+}
+
+export interface Rate {
+  requestRate: number;
+  errorRate: number;
+  errorRatio: number;
+}
+
+export const regexProtocol = {
+  grpc: new RegExp(/^[1-9]$|^1[0-6]$/),
+  /* eslint-disable-next-line */
+  http: new RegExp(/^[4-5]\d\d$/)
+};
+
+const getConfig = (ns: string, name: string, kind: string): RateHealthConfig | undefined => {
+  if (serverConfig.healthConfig && serverConfig.healthConfig.rate) {
+    for (let rate of serverConfig.healthConfig.rate) {
+      if (rate.namespace.test(ns) && rate.name.test(name) && rate.kind.test(kind)) {
+        return rate;
+      }
+    }
+  }
+  return undefined;
+};
+
+export const calculateErrorRate = (
+  ns: string,
+  name: string,
+  kind: string,
+  requests: RequestHealth
+): { errorRatio: ErrorRatio; config: RateHealthConfig | undefined } => {
+  const conf = getConfig(ns, name, kind);
+  const inboundTolerances = conf?.tolerance.filter(tol => tol.direction.test('inbound'));
+  const inbound = aggregate(requests.inbound, inboundTolerances);
+  const outboundTolerances = conf?.tolerance.filter(tol => tol.direction.test('outbound'));
+  const outbound = aggregate(requests.outbound, outboundTolerances);
+  const global = calculateGlobal(inbound, outbound);
+
+  return {
+    errorRatio: {
+      global: calculateStatus(global),
+      inbound: calculateStatus(inbound, inboundTolerances),
+      outbound: calculateStatus(outbound, outboundTolerances)
+    },
+    config: conf
+  };
+};
+
+const calculateStatus = (
+  requests: { [key: string]: Rate },
+  tolerances?: ToleranceConfig[]
+): { status: ThresholdStatus; protocol: string } => {
+  let result: { status: ThresholdStatus; protocol: string } = {
+    status: {
+      value: RATIO_NA,
+      status: NA
+    },
+    protocol: ''
+  };
+  for (let [protocol, rate] of Object.entries(requests)) {
+    const auxStatus = getRequestErrorsStatus(
+      rate.errorRatio,
+      tolerances ? tolerances.filter(tol => tol.protocol.test(protocol))[0] : undefined
```

Comment:
It doesn't look correct to pick only the first tolerance that matches the protocol. There can be several tolerances defined per protocol, each targeting different codes (it was the scenario identified for this task: being able to make a distinction between 4xx and 5xx errors).

(I'm trying to think a little bit how we could fix that)

Reply:
I think these two functions `calculateStatus` and `aggregate` need somehow to be merged in a single one to fix the two problems I mentioned (to handle mulitple configs per protocol, and to compute error rate independently from tolerances)

Also, I'm not totally sure, but perhaps it will be easier to iterate through all tolerances, and for each one compute their health and produce a status. Instinctively I would proceed that way, but not sure where it would lead.

Reply:
I'm not the writter, but not sure how complex we want to go here.
To me, feature wise I understand that I may have a config to accept some degree of 400 error but I may want to be notified in a very first 500.
So, I guess that first matches should mark the status, but I guess @jotak is thinking to compose all ?

I agree that some example would help to define this.

Reply:
We talked about this...and we decide to use the first match...:(. I have not problem to adapt the code but please discuss this

Reply:
I understand
```yaml
Tolerance: []Tolerance{
						{
							Code:      "402",
							Protocol:  "http",
							Direction: ".*",
							Degraded:  0.1,
							Failure:   20,
						},
						{
							Code:      "500",
							Protocol:  "http",
							Direction: ".*",
							Degraded:  0.2,
							Failure:   30,
						},
					},
```

In that case we should save all ErrorRatio, Requestratio etc...for each Tolerance , and check which one have the worst status. Is this what you mean @jotak ¿

Reply:
Yes, first match was for a complete match, code + protocol, not just protocol, else it doesn't make sense.

> In that case we should save all ErrorRatio, Requestratio etc...for each Tolerance , and check which one have the worst status. Is this what you mean @jotak ¿

Yes, that's the idea. Not sure if it's better than the current approach, but I've the feeling it will simplify things, don't you think? We can discuss about it.